content/en/blog/_posts/2024-09-25-swap-ga-graduation/2024-09-25-swap-ga-graduation.md

```diff
@@ -0,0 +1,445 @@
+* `containerMemoryRequest`: The container's memory request.
+
+Swap limitation is configured as:
+`(containerMemoryRequest / nodeTotalMemory) × totalPodsSwapAvailable`
+
+In other words, the amount of swap that a container is able to use is proportionate to its
+memory request, the node's total physical memory and the total amount of swap memory on
+the node that is available for use by Pods.
+
+It is important to note that, for containers within Burstable QoS Pods, it is possible to
+opt-out of swap usage by specifying memory requests that are equal to memory limits.
+Containers configured in this manner will not have access to swap memory.
+
+## How does it work?
+
+There are a number of possible ways that one could envision swap use on a node.
+When swap is already provisioned and available on a node,
+the kubelet is able to be configured so that:
+- It can start with swap on.
+- It will direct the Container Runtime Interface to allocate zero swap memory
+  to Kubernetes workloads by default.
+
+Swap configuration on a node is exposed to a cluster admin via the
+[`memorySwap` in the KubeletConfiguration](/docs/reference/config-api/kubelet-config.v1).
+As a cluster administrator, you can specify the node's behaviour in the
+presence of swap memory by setting `memorySwap.swapBehavior`.
+
+The kubelet employs the [CRI](https://kubernetes.io/docs/concepts/architecture/cri/)
+(container runtime interface) API, and directs the container runtime to
+configure specific cgroup v2 parameters (such as `memory.swap.max`) in a manner that will
+enable the desired swap configuration for a container. For runtimes that use control groups,
+the container runtime is then responsible for writing these settings to the container-level cgroup.
+
+## How can I monitor swap?
+
+### Node-level metric statistics
+
+Kubelet now collects
+[node-level metric statistics](/docs/reference/instrumentation/node-metrics/),
+which can be accessed at the `/metrics/resource` (which is used mainly by monitoring
+tools like Prometheus) and `/stats/summary` (which is used mainly by Autoscalers) kubelet HTTP endpoints.
+This allows clients who can directly interrogate the kubelet to
+monitor swap usage and remaining swap memory when using `LimitedSwap`.
+Additionally, a `machine_swap_bytes` metric has been added to cadvisor to show
+the total physical swap capacity of the machine.
+
+### NFD (Node Feature Discovery)
+
+[Node Feature Discovery](https://github.com/kubernetes-sigs/node-feature-discovery)
+is a Kubernetes addon for detecting hardware features and configuration.
+It can be utilized to discover which nodes are provisioned with swap.
+
+As an example, to figure out if a certain node is provisioned with swap,
+the following command can be executed while replacing `<NODE-NAME>` with the node's name:
+```shell
+> k get node <NODE-NAME> -o yaml | grep "feature.node.kubernetes.io/memory-swap"
```

Comment:
If we can use `jsonpath` output format rather than rely on `grep`, even better. This is informal but I'm minded to teach something that's 100% reliable. People might infer that this code is what they should put in scripts.

Also, what if the node is ever annotated `feature.node.kubernetes.io/memory-swap: "false"`?